src/ripple/overlay/Slot.h

```diff
@@ -0,0 +1,607 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_OVERLAY_SLOT_H_INCLUDED
+#define RIPPLE_OVERLAY_SLOT_H_INCLUDED
+
+#include <ripple/overlay/Squelch.h>
+#include <ripple/overlay/SquelchCommon.h>
+#include <ripple/protocol/PublicKey.h>
+#include <ripple.pb.h>
+
+namespace ripple {
+
+namespace Squelch {
+
+template <typename Peer, typename clock_type>
+class Slots;
+
+/** Peer's State */
+enum class PeerState : uint8_t {
+    Counting = 0x01,   // counting messages
+    Selected = 0x02,   // selected to relay, counting if Slot in Counting
+    Squelched = 0x03,  // squelched, doesn't relay
+};
+/** Slot's State */
+enum class SlotState : uint8_t {
+    Counting = 0x01,  // counting messages
+    Selected = 0x02,  // peers selected, stop counting
+};
+
+template <typename Unit, typename TP>
+Unit
+epoch(TP const& t)
+{
+    return duration_cast<Unit>(t.time_since_epoch());
+}
+
+/**
+ * Slot is associated with a specific validator via validator's public key.
+ * Slot counts messages from a validator, selects peers to be the source
+ * of the messages, and communicates the peers to be squelched. Slot can be
+ * in the following states: 1) Counting. This is the peer selection state
+ * when Slot counts the messages and selects the peers; 2) Selected. Slot
+ * doesn't count messages in this state but a received message
+ * may transition Slot to Counting state.
+ */
+template <typename Peer, typename clock_type>
```

Comment:
I understand the motivation to make this class generic - it help with testing. However, making this generic makes the code much harder to navigate and understand. We really need to think hard if there's another way to test the code that doesn't necessitate making this class generic.

This is not just a nit to me. Making this class use the concrete peer type, getting rid of all the generic callback functions, and getting rid of the comparator parameter would be a huge improvement in code understandably.

I do know we did something similar in the consensus code. Please think hard if there's another, better solution to the testing problem;

Reply:
We can  implement `Peer` class in unit tests with most of the methods simply returning some value. This is not going to be the test framework or a step towards implementing the test framework but a way to avoid  the extra allocation in (4). We can then pass `std::weak_ptr<Peer>` to all applicable slot methods as we do today with `Peer` being a concrete type so no template is required. We will have abstract class `SquelchHandler` with `squelch` and `unsquelch` methods. The class is going to be implemented in `Overlay` and `UnitTest`. Reference to the base class is going to be passed to `Slots` and `Slot` constructors and `squelch` and `unsquelch` called as needed. `UnitTest` can overwrite the handling on a fly providing flexibility of different handlers as needed. Does this work?